### PR TITLE
[AutoDiff upstream] Add derivative function type calculation.

### PR DESCRIPTION
**Pull request by dan-zheng**
Add `AnyFunctionType::getAutoDiffDerivativeFunctionType`, which returns the
derivative `AnyFunctionType` for an "original" `AnyFunctionType`, given:
- Differentiability parameter indices
- Differentiability result index
- Derivative function kind
- Derivative function generic signature (optional)
- Other auxiliary parameters

Add doc comments explaining typing rules, preconditions, and other details.

---

Progress towards TF-828: upstream `@differentiable` attribute type-checking.
`@differentiable` attribute type-checking mega-patch: https://github.com/apple/swift/pull/29091

---

Tests exist on `tensorflow` branch and will be upstreamed later.
Currently, not enough code has been upstreamed for meaningful testing.

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

**Comment by dan-zheng** [comment]
macOS flaky test failed (passes locally)
```
03:59:27 Failing Tests (1):
03:59:27     Swift(macosx-x86_64) :: ModuleInterface/ModuleCache/prebuilt-module-cache-forwarding.swift
```
Rerunning CI now.

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test macOS platform

**Comment by dan-zheng** [comment]
Merging to unblock progress. Happy to address review feedback later.

File: lib/AST/Type.cpp [hunk 1]
```diff
@@ -4910,6 +4910,134 @@ TypeBase::getAutoDiffTangentSpace(LookupConformanceFn lookupConformance) {
...
+          diffParamType->getAutoDiffTangentSpace(lookupConformance)
+              ->getType()));
+    SmallVector<TupleTypeElt, 8> differentialResults;
+    if (auto *resultTuple = originalResult->getAs<TupleType>()) {
+      auto resultTupleEltType = resultTuple->getElementType(resultIndex);
+      differentialResults.push_back(
+          resultTupleEltType->getAutoDiffTangentSpace(lookupConformance)
+              ->getType());
+    } else {
+      assert(resultIndex == 0 && "resultIndex out of bounds");
```
Comment (dan-zheng) [inline lib/AST/Type.cpp]: Minor: this should match the assertion below:
```
      assert(resultIndex == 0 &&
             "Expected result index 0 for non-tuple result");
```

File: lib/AST/Type.cpp [hunk 2]
```diff
@@ -4910,6 +4910,134 @@ TypeBase::getAutoDiffTangentSpace(LookupConformanceFn lookupConformance) {
...
+    break;
+  }
+  }
+  assert(linearMapType && "Expected linear map type");
+
+  // Build the full derivative function type: `(T...) -> (R, LinearMapType)`.
+  SmallVector<TupleTypeElt, 2> retElts;
+  retElts.push_back(originalResult);
+  retElts.push_back(linearMapType);
+  auto retTy = TupleType::get(retElts, ctx);
```
Comment (dan-zheng) [inline lib/AST/Type.cpp]: Minor: I prefer using the name "result type" instead of "return type". `retTy` -> `resultTy`

File: lib/AST/Type.cpp [hunk 3]
```diff
@@ -4910,6 +4910,134 @@ TypeBase::getAutoDiffTangentSpace(LookupConformanceFn lookupConformance) {
   return cache(None);
 }
 
+// Creates an `AnyFunctionType` from the given parameters, result type,
+// generic signature, and `ExtInfo`.
+static AnyFunctionType *
+makeFunctionType(ArrayRef<AnyFunctionType::Param> parameters, Type resultType,
```
Comment (rxwei) [inline lib/AST/Type.cpp]: This is a general utility for AST function types. Consider adding it to `AnyFunctionType`, or just making it a closure inside `getAutoDiffDerivativeFunctionType`.

**Pull request merged**